File: programmable_video/ios/Classes/RemoteDataTrackListener.swift
import Flutter
import Foundation
import TwilioVideo

class RemoteDataTrackListener: BaseListener, RemoteDataTrackDelegate {
    let TAG = "RemoteDataTrackListener"

    func remoteDataTrackDidReceiveData(remoteDataTrack: RemoteDataTrack, message: Data) {
        debug("didReceiveData => " +
            "sid: \(remoteDataTrack.sid), " +
            "message: \(message)"
        )
        sendEvent("bufferMessage", data: [
            "remoteDataTrack": RemoteDataTrackListener.remoteDataTrackToDict(remoteDataTrack) as Any,
            "message": message
        ])
    }

    func remoteDataTrackDidReceiveString(remoteDataTrack: RemoteDataTrack, message: String) {
        debug("didReceiveString => " +
            "sid: \(remoteDataTrack.sid), " +
            "message: \(message)"
        )
        sendEvent("stringMessage", data: [
            "remoteDataTrack": RemoteDataTrackListener.remoteDataTrackToDict(remoteDataTrack) as Any,
            "message": message
        ])
    }

    static func remoteDataTrackToDict(_ remoteDataTrack: RemoteDataTrack?) -> [String: Any]? {
        if let remoteDataTrack = remoteDataTrack {
            return [
                "sid": remoteDataTrack.sid,
                "name": remoteDataTrack.name,
                "enabled": remoteDataTrack.isEnabled,
                "ordered": remoteDataTrack.isOrdered,
                "reliable": remoteDataTrack.isReliable,
                "maxPacketLifeTime": remoteDataTrack.maxPacketLifeTime,
                "maxRetransmits": remoteDataTrack.maxRetransmits
            ]
        }
        return nil
    }

    func debug(_ msg: String) {
        SwiftTwilioProgrammableVideoPlugin.debug("\(TAG)::\(msg)")
    }
}

File: programmable_video/android/src/main/kotlin/twilio/flutter/twilio_programmable_video/RemoteDataTrackListener.kt
package twilio.flutter.twilio_programmable_video

import com.twilio.video.RemoteDataTrack
import java.nio.ByteBuffer

class RemoteDataTrackListener : BaseListener(), RemoteDataTrack.Listener {
    private val TAG = "RemoteDataTrackListener"

    override fun onMessage(remoteDataTrack: RemoteDataTrack, message: ByteBuffer) {
        debug("onMessage => sid: ${remoteDataTrack.sid}, message (ByteBuffer): $message ")

        sendEvent("bufferMessage", mapOf(
                "remoteDataTrack" to remoteDataTrackToMap(remoteDataTrack),
                "message" to message.array()
        ))
    }

    override fun onMessage(remoteDataTrack: RemoteDataTrack, message: String) {
        debug("onMessage => sid: ${remoteDataTrack.sid}, message (String): $message")

        sendEvent("stringMessage", mapOf(
                "remoteDataTrack" to remoteDataTrackToMap(remoteDataTrack),
                "message" to message
        ))
    }

    companion object {
        @JvmStatic
        fun remoteDataTrackToMap(remoteDataTrack: RemoteDataTrack?): Map<String, Any>? {
            if (remoteDataTrack != null) {
                return mapOf(
                        "sid" to remoteDataTrack.sid,
                        "name" to remoteDataTrack.name,
                        "enabled" to remoteDataTrack.isEnabled,
                        "ordered" to remoteDataTrack.isOrdered,
                        "reliable" to remoteDataTrack.isReliable,
                        "maxPacketLifeTime" to remoteDataTrack.maxPacketLifeTime,
                        "maxRetransmits" to remoteDataTrack.maxRetransmits
                )
            }
            return null
        }
    }

    internal fun debug(msg: String) {
        TwilioProgrammableVideoPlugin.debug("$TAG::$msg")
    }
}
